### PR TITLE
Before/After middleware can be set via yaml/xml and supported HTTP methods extended

### PR DESCRIPTION
**Pull request by samir-plusb**
Two commits:

- The first commit adds the possibility to specify a before/after route middleware class
- The second commit adds the missing HTTP-request methods. Now the methods 'get', 'put', 'post', 'delete', 'options', 'head', 'patch', 'purge', 'options', 'trace', 'connect' are supported

### CODE REVIEW

File: src/MJanssen/Provider/RoutingServiceProvider.php [hunk 1]
```diff
@@ -6,13 +6,13 @@
 use Silex\Controller;
 use Silex\Route;
 use Silex\ServiceProviderInterface;
+use Symfony\Component\VarDumper\VarDumper;
```
Comment (marcojanssen) [inline src/MJanssen/Provider/RoutingServiceProvider.php]: Not used

File: src/MJanssen/Provider/RoutingServiceProvider.php [hunk 2]
```diff
@@ -6,13 +6,13 @@
...
 
 /**
  * Class RoutingServiceProvider
  * @package MJanssen\Provider
  */
-class RoutingServiceProvider implements ServiceProviderInterface
-{
+class RoutingServiceProvider implements ServiceProviderInterface {
```
Comment (marcojanssen) [inline src/MJanssen/Provider/RoutingServiceProvider.php]: The braces should be on the next line according to PSR

Comment (marcojanssen) [inline src/MJanssen/Provider/RoutingServiceProvider.php]: Did your editor auto correct it?

File: src/MJanssen/Provider/RoutingServiceProvider.php [hunk 3]
```diff
@@ -212,8 +204,66 @@ protected function addActions(Controller $controller, $actions, $type)
...
-    protected function addAction(Controller $controller, $name, $value, $type)
-    {
+    protected function addAction(Controller $controller, $name, $value, $type) {
         call_user_func_array(array($controller, $type), array($name, $value));
     }
+
+    /**
+     * Determines whether a variable is valid Closure
+     *
+     * source: http://stackoverflow.com/questions/7101469/determining-if-a-variable-is-a-valid-closure-in-php
```
Comment (marcojanssen) [inline src/MJanssen/Provider/RoutingServiceProvider.php]: Why mention the source?

File: src/MJanssen/Provider/RoutingServiceProvider.php [hunk 4]
```diff
@@ -212,8 +204,66 @@ protected function addActions(Controller $controller, $actions, $type)
...
      * @param $type
      */
-    protected function addAction(Controller $controller, $name, $value, $type)
-    {
+    protected function addAction(Controller $controller, $name, $value, $type) {
         call_user_func_array(array($controller, $type), array($name, $value));
     }
+
+    /**
+     * Determines whether a variable is valid Closure
```
Comment (marcojanssen) [inline src/MJanssen/Provider/RoutingServiceProvider.php]: If we need a description what the method does then our method name isn't clear enough:

```
isClosure($param)
```

Seems to me that the method name is fine as is, it's already saying what it does

File: src/MJanssen/Provider/RoutingServiceProvider.php [hunk 5]
```diff
@@ -212,8 +204,66 @@ protected function addActions(Controller $controller, $actions, $type)
...
+            );
+        }
+
+        if ($this->isClosure($value)) {
+            //When a closure is provided, we will just load it as a middleware type
+            $controller->$type($value);
+        } else {
+            //In this case a yaml/xml configuration was used
+
+            if (!is_string($value) || strpos($value, '::') === FALSE) {
```
Comment (marcojanssen) [inline src/MJanssen/Provider/RoutingServiceProvider.php]: You should extract this part and give it a good name so the intent is clear

```
if ($this->isClosure($value)) {
   // do stuff
} else {
    $this->addMiddlewareFromConfig();
}
```

File: tests/MJanssen/Provider/RoutingServiceProviderTest.php [hunk 6]
```diff
@@ -389,4 +390,29 @@ public function testInvalidBefore()
         $routingServiceProvider->addRoute($app, $route);
 
     }
+
+    /**
+     *
+     */
+    public function testAddBeforeAfterMiddlewareByString()
```
Comment (marcojanssen) [inline tests/MJanssen/Provider/RoutingServiceProviderTest.php]: What are we testing exactly? There are no asserts or spies

File: tests/MJanssen/Provider/RoutingServiceProviderTest.php [hunk 7]
```diff
@@ -389,4 +390,29 @@ public function testInvalidBefore()
...
+    {
+        $app = new Application();
+        $routingServiceProvider = new RoutingServiceProvider();
+        $route = $this->validRoute;
+
+        $mwClassName = 'FooMiddleware';
```
Comment (marcojanssen) [inline tests/MJanssen/Provider/RoutingServiceProviderTest.php]: Better to write it in full, $middlewareClassName, abbreviations just confuse people.

File: src/MJanssen/Provider/RoutingServiceProvider.php [hunk 8]
```diff
@@ -212,8 +204,66 @@ protected function addActions(Controller $controller, $actions, $type)
...
+                    )
+                );
+            }
+
+            list($class, $method) = explode('::', $value, 2);
+
+            if ($class && $method) {
+
+                if (!method_exists($class, $method)) {
+                    throw new \BadMethodCallException(sprintf('Method "%s::%s" does not exist.', $class, $method));
```
Comment (marcojanssen) [inline src/MJanssen/Provider/RoutingServiceProvider.php]: All these exceptions are test cases and all are not tested

**Comment by samir-plusb** [comment]
Hey Marco,
thank you for the feedback. I implemented you suggestions and will make a new pull request.

**Pull request closed without merging**